django/core/paginator.py

```diff
@@ -178,6 +183,140 @@ def get_elided_page_range(self, number=1, *, on_each_side=3, on_ends=2):
             yield from range(number + 1, self.num_pages + 1)
 
 
+class AsyncPaginator(BasePaginator):
+    def __init__(
+        self,
+        object_list,
+        per_page,
+        orphans=0,
+        allow_empty_first_page=True,
+        error_messages=None,
+    ):
+        super().__init__(
+            object_list, per_page, orphans, allow_empty_first_page, error_messages
+        )
+        self._cache_acount = None
+        self._cache_anum_pages = None
+
+    async def __aiter__(self):
+        page_range = await self.apage_range()
+        for page_number in page_range:
+            yield await self.apage(page_number)
+
+    async def avalidate_number(self, number):
+        """See Paginator.validate_number()."""
+        try:
+            if isinstance(number, float) and not number.is_integer():
+                raise ValueError
+            number = int(number)
+        except (TypeError, ValueError):
+            raise PageNotAnInteger(self.error_messages["invalid_page"])
+        if number < 1:
+            raise EmptyPage(self.error_messages["min_page"])
+        num_pages = await self.anum_pages()
+        if number > num_pages:
+            raise EmptyPage(self.error_messages["no_results"])
+        return number
```

Comment:
I think we can add to the base class
```python
    def _validate_number(self, number, num_pages):
        """Validate the given 1-based page number."""
        try:
            if isinstance(number, float) and not number.is_integer():
                raise ValueError
            number = int(number)
        except (TypeError, ValueError):
            raise PageNotAnInteger(self.error_messages["invalid_page"])
        if number < 1:
            raise EmptyPage(self.error_messages["min_page"])
        if number > num_pages:
            raise EmptyPage(self.error_messages["no_results"])
        return number
```
and then update this to 
```suggestion
    async def avalidate_number(self, number):
        num_pages = await self.anum_pages()
        return self._validate_number(number, num_pages=num_pages)
```
and the other to
```python
    def validate_number(self, number):
        return self._validate_number(number, num_pages=self.num_pages)
```